File: go/sketchkit/hash/hash.go
// Code authors: Vijay Erramilli and Codex
// Package hash implements domain-separated keyed hashing for sketch inputs.
package hash

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"fmt"
	"os"
)

// Algorithm names a hash algorithm from spec/hash.md.
type Algorithm string

const (
	// HMACSHA25664 is HMAC-SHA256 truncated to the first 64 bits.
	HMACSHA25664 Algorithm = "hmac_sha256_64"

	minSecretBytes = 16
)

// Domain names a registered hash domain from spec/hash.md.
type Domain string

const (
	PromptV1       Domain = "prompt:v1"
	UserV1         Domain = "user:v1"
	ToolV1         Domain = "tool:v1"
	RetrievalDocV1 Domain = "retrieval-doc:v1"
	SessionV1      Domain = "session:v1"
	MCPSessionV1   Domain = "mcp-session:v1"
	MCPMethodV1    Domain = "mcp-method:v1"
	ToolErrorV1    Domain = "tool-error:v1"
)

var registeredDomains = map[Domain]struct{}{
	PromptV1:       {},
	UserV1:         {},
	ToolV1:         {},
	RetrievalDocV1: {},
	SessionV1:      {},
	MCPSessionV1:   {},
	MCPMethodV1:    {},
	ToolErrorV1:    {},
}

var domainList = []Domain{
	PromptV1,
	UserV1,
	ToolV1,
	RetrievalDocV1,
	SessionV1,
	MCPSessionV1,
	MCPMethodV1,
	ToolErrorV1,
}

var (
	// ErrEmptySecret reports that no hash secret was supplied.
	ErrEmptySecret = errors.New("empty hash secret")

	// ErrWeakSecret reports secret material that is too short or is a known placeholder.
	ErrWeakSecret = errors.New("weak hash secret")

	// ErrEmptySecretEnv reports that no environment variable name was supplied.
	ErrEmptySecretEnv = errors.New("empty hash secret environment variable")

	// ErrUnregisteredDomain reports a domain that is absent from the registry.
	ErrUnregisteredDomain = errors.New("unregistered hash domain")
)

// Secret is opaque keyed-hash secret material.
type Secret struct {
	value []byte
}

// SecretFromEnv loads opaque secret bytes from envName.
func SecretFromEnv(envName string) (Secret, error) {
	if envName == "" {
		return Secret{}, ErrEmptySecretEnv
	}

	value := os.Getenv(envName)
	if value == "" {
		return Secret{}, fmt.Errorf("%w: %s", ErrEmptySecret, envName)
	}
	if err := validateSecretBytes([]byte(value)); err != nil {
		return Secret{}, err
	}

	return Secret{value: []byte(value)}, nil
}

// String redacts secret material in ordinary formatting.
func (Secret) String() string {
	return "<redacted hash secret>"
}

// GoString redacts secret material in %#v formatting.
func (Secret) GoString() string {
	return "hash.Secret(<redacted>)"
}

// Domains returns the registered v0.1 hash domains.
func Domains() []Domain {
	domains := make([]Domain, len(domainList))
	copy(domains, domainList)

	return domains
}

// IsRegisteredDomain reports whether domain is registered in spec/hash.md.
func IsRegisteredDomain(domain Domain) bool {
	_, ok := registeredDomains[domain]
	return ok
}

// Hash64 returns HMAC-SHA256(secret, domain || 0x00 || canonicalBytes) truncated to uint64.
func Hash64(secret Secret, domain Domain, canonicalBytes []byte) (uint64, error) {
	digest, err := Digest64(secret, domain, canonicalBytes)
	if err != nil {
		return 0, err
	}

	return binary.BigEndian.Uint64(digest[:]), nil
}

// Digest64 returns the first eight HMAC-SHA256 digest bytes.
func Digest64(secret Secret, domain Domain, canonicalBytes []byte) ([8]byte, error) {
	if err := validateSecretBytes(secret.value); err != nil {
		return [8]byte{}, err
	}
	if !IsRegisteredDomain(domain) {
		return [8]byte{}, fmt.Errorf("%w: %s", ErrUnregisteredDomain, domain)
	}

	mac := hmac.New(sha256.New, secret.value)
	mac.Write([]byte(domain))
	mac.Write([]byte{0x00})
	mac.Write(canonicalBytes)

	sum := mac.Sum(nil)
	var digest [8]byte
	copy(digest[:], sum[:8])

	return digest, nil
}

func validateSecretBytes(value []byte) error {
	if len(value) == 0 {
		return ErrEmptySecret
	}
	if len(value) < minSecretBytes {
		return fmt.Errorf("%w: must be at least %d bytes", ErrWeakSecret, minSecretBytes)
	}
	switch string(value) {
	case "demo-secret-change-me", "dev-secret-change-me":
		return fmt.Errorf("%w: placeholder must be replaced", ErrWeakSecret)
	}
	return nil
}

// Digest64Hex returns the lowercase fixed-width hexadecimal digest.
func Digest64Hex(secret Secret, domain Domain, canonicalBytes []byte) (string, error) {
	digest, err := Digest64(secret, domain, canonicalBytes)
	if err != nil {
		return "", err
	}

	return hex.EncodeToString(digest[:]), nil
}

File: python/llm_sketchkit/hash.py
# Code authors: Vijay Erramilli and Codex
"""Domain-separated keyed hashing for sketch inputs."""

from __future__ import annotations

import hashlib
import hmac
import os
from dataclasses import dataclass

from . import profiles

Algorithm = str
Domain = str

HMAC_SHA256_64 = profiles.HMAC_SHA256_64
MIN_SECRET_BYTES = 16
PROMPT_V1 = profiles.PROMPT_V1
USER_V1 = profiles.USER_V1
TOOL_V1 = profiles.TOOL_V1
RETRIEVAL_DOC_V1 = profiles.RETRIEVAL_DOC_V1
SESSION_V1 = profiles.SESSION_V1
MCP_SESSION_V1 = profiles.MCP_SESSION_V1
MCP_METHOD_V1 = profiles.MCP_METHOD_V1
TOOL_ERROR_V1 = profiles.TOOL_ERROR_V1


class HashError(ValueError):
    """Base class for keyed-hash failures."""


class EmptySecretError(HashError):
    """Raised when no secret material is supplied."""


class EmptySecretEnvError(HashError):
    """Raised when no secret environment variable name is supplied."""


class WeakSecretError(HashError):
    """Raised when secret material is too short or a known placeholder."""


class UnregisteredDomainError(HashError):
    """Raised when a domain is not in the v0.1 registry."""


@dataclass(frozen=True, repr=False)
class Secret:
    """Opaque keyed-hash secret material."""

    _value: bytes

    def __post_init__(self) -> None:
        _validate_secret(self._value)

    def __repr__(self) -> str:
        return "Secret(<redacted>)"

    def __str__(self) -> str:
        return "<redacted hash secret>"


def secret_from_env(env_name: str) -> Secret:
    """Load opaque secret bytes from an environment variable."""

    if not env_name:
        raise EmptySecretEnvError("empty hash secret environment variable")
    value = os.environ.get(env_name)
    if not value:
        raise EmptySecretError(env_name)
    return Secret(value.encode("utf-8"))


def domains() -> list[Domain]:
    """Return registered v0.1 hash domains in spec order."""

    return [
        PROMPT_V1,
        USER_V1,
        TOOL_V1,
        RETRIEVAL_DOC_V1,
        SESSION_V1,
        MCP_SESSION_V1,
        MCP_METHOD_V1,
        TOOL_ERROR_V1,
    ]


def is_registered_domain(domain: Domain) -> bool:
    """Return whether a hash domain is registered."""

    return domain in profiles.REGISTERED_DOMAINS


def digest64(secret: Secret, domain: Domain, canonical_bytes: bytes) -> bytes:
    """Return the first eight HMAC-SHA256 digest bytes."""

    _validate_secret(secret._value)
    if not is_registered_domain(domain):
        raise UnregisteredDomainError(domain)
    message = domain.encode("ascii") + b"\x00" + canonical_bytes
    return hmac.new(secret._value, message, hashlib.sha256).digest()[:8]


def hash64(secret: Secret, domain: Domain, canonical_bytes: bytes) -> int:
    """Return HMAC-SHA256-64 as an unsigned big-endian integer."""

    return int.from_bytes(digest64(secret, domain, canonical_bytes), "big")


def digest64_hex(secret: Secret, domain: Domain, canonical_bytes: bytes) -> str:
    """Return the lowercase fixed-width hexadecimal digest."""

    return digest64(secret, domain, canonical_bytes).hex()


def _validate_secret(value: bytes) -> None:
    if not value:
        raise EmptySecretError("empty hash secret")
    if len(value) < MIN_SECRET_BYTES:
        raise WeakSecretError(f"hash secret must be at least {MIN_SECRET_BYTES} bytes")
    if value in {b"demo-secret-change-me", b"dev-secret-change-me"}:
        raise WeakSecretError("placeholder secret must be replaced")
